Gems/ROS2/Code/Source/Manipulator/JointPublisherComponent.cpp

```diff
@@ -0,0 +1,148 @@
+#include <Source/HingeJointComponent.h>
+#include <AzCore/Component/ComponentApplicationBus.h>
+#include <ROS2/Frame/ROS2FrameComponent.h>
+#include <ROS2/Utilities/ROS2Names.h>
+
+namespace ROS2
+{
+    void JointPublisherComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+        auto ros2Node = ROS2::ROS2Interface::Get()->GetNode();
+        auto ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
+        AZStd::string namespacedTopic = ROS2Names::GetNamespacedName(ros2Frame->GetNamespace(), "joint_states");
+        m_jointstatePublisher = ros2Node->create_publisher<sensor_msgs::msg::JointState>(namespacedTopic.data(), rclcpp::QoS(1));        // TODO: add QoS instead of "1"
+    }
+
+    void JointPublisherComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+        m_jointstatePublisher.reset();
+    }
+
+    void JointPublisherComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        provided.push_back(AZ_CRC_CE("JointPublisher"));
+    }
+
+
+    void JointPublisherComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<JointPublisherComponent, AZ::Component>()
+                ->Version(0)
+                ->Field("Frequency (HZ)", &JointPublisherComponent::m_frequency);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<JointPublisherComponent>("JointPublisherComponent", "[Publish all the Hinge joint in the tree]")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, &JointPublisherComponent::m_frequency, "Frequency", "Frequency of publishing [Hz]");
+            }
+        }
+    }
+
+    void JointPublisherComponent::InitializeMap()
+    {
+        AZStd::vector<AZ::EntityId> descendants;
+        AZ::TransformBus::EventResult(descendants, GetEntityId(), &AZ::TransformInterface::GetAllDescendants);
+
+        for (const AZ::EntityId& descendantID : descendants)
+        {
+            AZ::Entity* entity = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationRequests::FindEntity, descendantID);
+            AZ_Assert(entity, "Unknown entity %s", descendantID.ToString().c_str());
+            auto* frameComponent = entity->FindComponent<ROS2FrameComponent>();
+            auto* hingeComponent = entity->FindComponent<PhysX::HingeJointComponent>();
+            if (frameComponent && hingeComponent)
+            {
+                m_hierarchyMap[frameComponent->GetJointName()] = descendantID;
```

Comment:
I am still building a map to store the entityID and the jointName at the launching of the Game mode. Any comments on how to avoid that or better/more efficient solutions?